urlbc3e98d97e_jmhorjus_EFCP_Implementation_tgz-pJ8-UDPConnector_UdpConnectorTestJ8/src/UDPConnector/EfcpConnector.java

```diff
@@ -0,0 +1,515 @@
+            }
+            catch(Exception ex) { 
+                System.out.print("Exception Sending Ack:" + ex.getMessage() + "\n"); 
+            }
+        }
+        
+        
+        void ReceiveAck(int sequenceNum)
+        {
+            System.out.print("Efcp: Ack Received: seq="+sequenceNum+"\n");
+            if (m_senderLastPacketAcked >= sequenceNum) {
+                System.out.print("Ignoring redundant Ack. lastAck="+m_senderLastPacketAcked+"\n");
+                return;
+            }
+            //1.) Cancel all retransmission events for packets with 
+            // sequence numbers <= the acked sequence number.
+            while(m_senderLastPacketAcked < sequenceNum)
+            {
+                ScheduledFuture task = m_senderRetransmitQueue.get(++m_senderLastPacketAcked);
+                task.cancel(false);
+            }
+        }
+        
+        
+        void ReceiveFlowControl(DtcpPacket packet)
+        {
+            if (m_senderRightWindowEdge < packet.NewRightWindowEdge) {
+                System.out.print("Efcp: Flow control received. right window edge from "
+                        + m_senderRightWindowEdge + " to " + packet.NewRightWindowEdge + ".\n");
+                // Set the new right window edge.
+                m_senderRightWindowEdge = packet.NewRightWindowEdge;
+            
+                // Rate based control:
+                if(m_policyInfo.RateFlowControlEnabled){
+                    m_policyInfo.RateDefaultPeriodInMs = packet.NewDataPeriodInMs;
+                    m_policyInfo.RateDefaultPaketsPerPeriod = packet.NewDataRate;
+                }
+                    
+            }
+            
+            //2.) If there are packet waiting on the closed window queue, 
+            // immediately send as many of them as the new window allows.
+            while(!m_senderClosedWindowQueue.isEmpty() &&
+                    m_senderRightWindowEdge > m_senderClosedWindowQueue.peek().getSeqNum() &&
+                    IsRateBasedFlowControlWindowOpen())
+            {
+                try { SendPacket(m_senderClosedWindowQueue.remove()); }
+                catch(Exception ex) { System.out.print("Exception Sending: "+ex.getMessage()+"\n"); }
+            }
+            
+        }
+        
+    }
+
+    class RatePeriodExpiredEvent implements Runnable 
+    {
+        Date m_scheduleTime;
+        RatePeriodExpiredEvent(Date now)
+        {
+            m_scheduleTime = now;
+        }
+        
+        @Override
+        public void run()
+        {
+            System.out.print("***Efcp: Rate based flow control period expired. Packets Waiting="
+                   + m_senderClosedWindowQueue.size() 
+                   + " windowEdge=" + m_senderRightWindowEdge+"\n");
+            
+            //1.) Reset the rate period data.
+            synchronized(m_senderRateCurrentPeriodStartTime)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Object m_senderRateCurrentPeriodStartTime synchronized on line 478 should not be re-assigned in the same synchronized block on line 497. After this re-assignment, you lose the reference to the synchronized object.